tools/privacyidea-migrate-linotp.py

```diff
@@ -173,182 +220,135 @@
+            resolver = ASSIGNMENTS.get("resolver").get(linotp_resolver)
+            resolver_type = resolver_type
+            user_id = r['LinOtpUserid']
+        else:
+            user_pin = None
+            user_pin_iv = None
+            resolver = None
+            resolver_type = None
+            user_id = None
+
+        token_values.append(dict(
+            id=r["LinOtpTokenId"],
+            description=r["LinOtpTokenDesc"],
+            serial=r["LinOtpTokenSerialnumber"],
+            tokentype=type,
+            user_pin=user_pin,
+            user_pin_iv=user_pin_iv,
+            so_pin=r['LinOtpTokenPinSO'],
+            so_pin_iv=r['LinOtpTokenPinSOIV'],
+            resolver=resolver,
+            resolver_type=resolver_type,
+            user_id=user_id,
+            pin_seed=r['LinOtpSeed'],
+            key_enc=r['LinOtpKeyEnc'],
+            key_iv=r['LinOtpKeyIV'],
+            maxfail=r['LinOtpMaxFail'],
+            active=r['LinOtpIsactive'],
+            failcount=r['LinOtpFailCount'],
+            count=r['LinOtpCount'],
+            count_window=r['LinOtpCountWindow'],
+            sync_window=r['LinOtpSyncWindow']))
+
+        if MIGRATE.get("tokeninfo") and ti:
+            # Add tokeninfo for this token
+            for k, v in ti.iteritems():
+                tokeninfo_values.append(dict(
+                    serial=r["LinOtpTokenSerialnumber"],
+                    Key=k, Value=v,
+                    token_id=r["LinOtpTokenId"]))
+                print(" +--- processing tokeninfo {0!s}".format(k))
+
+    print
+    print("Adding {} tokens...".format(len(token_values)))
+    conn_pi.execute(token_table.insert(), token_values)
+
+    if MIGRATE.get("tokeninfo"):
+        # fet the new token_id's in privacyIDEA
+        s = select([token_table])
         result = conn_pi.execute(s)
         for r in result:
-            resolver_id = r["id"]
-        # insert each resolver_id for the realm_id
-        ins = resolverrealm_table.insert().values(
-            realm_id=realm_id,
-            resolver_id=resolver_id
-        )
-        print(" +--- Adding resolver {0!s} to realm {1!s}".format(
-            resolvername, realmname))
-        conn_pi.execute(ins)
+            token_serial_id_map[r["serial"]] = r["id"]
+
+        # Now we have to rewrite the token_id in the tokeninfo_values
+        for ti in tokeninfo_values:
+            ti["token_id"] = token_serial_id_map[ti["serial"]]
+            del ti["serial"]
+
+        print("Adding {} token infos...".format(len(tokeninfo_values)))
+        conn_pi.execute(tokeninfo_table.insert(), tokeninfo_values)
+
+if MIGRATE.get("assignments") and resolver:
+    # If the token is assigned, we also need to create an entry for tokenrealm
+    # We need to determine the realm_id for this resolver!
+    for token in token_values:
+        token_id = token.get("id")
+        resolver = token.get("resolver")
+        realm = ASSIGNMENTS.get("realm").get(resolver)
```

Comment:
Maybe also check for ``realm is None`` here? (see above)

Reply:
s.a.